src/Nethermind/Nethermind.Consensus/Processing/CensorshipDetector/CensorshipDetector.cs

```diff
@@ -0,0 +1,228 @@
+            {
+                if (Address.TryParse(hexString, out Address address))
+                {
+                    _poolAddressCensorshipBestTx[address!] = null;
+                }
+            }
+        }
+
+        _potentiallyCensoredBlocks = new(_cacheSize, _cacheSize, "potentiallyCensoredBlocks");
+        _censoredBlocks = new(_cacheSize);
+        _blockProcessor.BlockProcessing += OnBlockProcessing;
+    }
+
+    private void OnBlockProcessing(object? sender, BlockEventArgs e)
+    {
+        UInt256 baseFee = e.Block.BaseFeePerGas;
+
+        IEnumerable<Transaction> poolBestTransactions = _txPool.GetBestTxOfEachSender();
+        foreach (Transaction tx in poolBestTransactions)
+        {
+            if (tx.To is not null && tx.GasBottleneck > baseFee && _poolAddressCensorshipBestTx.TryGetValue(tx.To!, out Transaction? bestTx))
+            {
+                if (bestTx is null)
+                {
+                    _poolAddressCensorshipBestTx[tx.To!] = tx;
+                }
+                else if (_comparer.Compare(bestTx, tx) > 0)
+                {
+                    _poolAddressCensorshipBestTx[tx.To!] = tx;
+                }
+            }
+        }
+
+        Task.Run(() => Cache(e.Block));
+    }
+
+    private void Cache(Block block)
+    {
+        Transaction bestTxInBlock = block.Transactions[0];
+        Transaction worstTxInBlock = block.Transactions[0];
+        HashSet<AddressAsKey> blockAddressCensorshipDetectorHelper = [];
+        // Number of unique addresses specified by the user for censorship detection, to which txs are sent in the block
+        long blockCensorshipDetectionUniqueAddressCount = 0;
+
+        foreach (Transaction tx in block.Transactions)
+        {
+            if (!tx.SupportsBlobs)
+            {
+                if (_poolAddressCensorshipBestTx.ContainsKey(tx.To!)
+                && !blockAddressCensorshipDetectorHelper.Contains(tx.To!))
+                {
+                    blockAddressCensorshipDetectorHelper.Add(tx.To!);
+                    blockCensorshipDetectionUniqueAddressCount++;
+                }
+
+                if (_comparer.Compare(bestTxInBlock, tx) > 0)
+                {
+                    bestTxInBlock = tx;
+                }
+                else
+                {
+                    worstTxInBlock = tx;
+                }
+            }
+        }
+
+        long poolCensorshipDetectionUniqueAddressCount = 0;
+        foreach (Transaction? bestTx in _poolAddressCensorshipBestTx.Values)
+        {
+            if (bestTx is not null && _comparer.Compare(bestTx, worstTxInBlock) <= 0)
+            {
+                poolCensorshipDetectionUniqueAddressCount++;
+            }
+        }
+
+        bool isCensored = _comparer.Compare(bestTxInBlock, _txPool.GetBestTx()) > 0
+        || blockCensorshipDetectionUniqueAddressCount * 2 < poolCensorshipDetectionUniqueAddressCount;
```

Comment:
What we have:
1. We are iterating TxPool txs (one - best one- per sender) and building `_poolAddressCensorshipBestTx`, which will have best tx which is sending to every receiver (so just best tx per every `To` pointed in config)
2. We are iterating all block txs and looking for receiver address (`tx.To`) in previously built `_poolAddressCensorshipBestTx`. If found, incrementing `blockCensorshipDetectionUniqueAddressCount`. So `blockCensorshipDetectionUniqueAddressCount` means a number of txs in the block which were sending txs to receiver `To` from the configured list. `blockAddressCensorshipDetectorHelper` is ensuring that we are counting every receiver only once.
3. We are iterating `_poolAddressCensorshipBestTx` and comparing with worst tx in the block. If tx in collection is better, incrementing `poolCensorshipDetectionUniqueAddressCount`. So `poolCensorshipDetectionUniqueAddressCount` means a number of configured receiver addresses with tx in our TxPool, which were better than worst tx in the block, but not included in the block
4. Then we are setting `isCensored` if at least one of 2 parts is true.
- comparing best tx in the block with best tx in our pool and `true` if tx from the pool is better. But we are checking just best tx, no matter which `To` it has. So we can have censorship detected even if we don't have in the pool any tx with `To` from the configured list. Is it expected?
- checking if 2 * number of receivers which have tx sending to them in the block is lower than number of receivers which have tx in our pool which is better than worst tx in the block.

In current form we will have censorship detected always, when we have in the pool ANY tx better than best tx in the block. Is it expected behavior? Maybe we could change `bool isCensored = _comparer.Compare(bestTxInBlock, _txPool.GetBestTx()) > 0` to check best tx from `_poolAddressCensorshipBestTx.Values` instead of `_txPool.GetBestTx()`? But I'm not sure what is expected behavior here

We are also considering that there is a censorship, if we found 2*X receivers with txs in our pool better than worst tx in the block, and in the block were included less than X receivers (ofc counting only receivers from the configured list)

Reply:
Probably it is fine as it is

Reply:
I don't think we need to change the isCensored method since we are dealing with two separate kinds of censorship detection here. I think it seems a bit convoluted because of a lack of comments.

Reply:
Ok this is the last line I have doubts about:
1. First part we are checking: `_comparer.Compare(bestTxInBlock, _txPool.GetBestTx()) > 0`, this means we found transaction in txPool that is better than best transaction in block - right? Should we quantify it in any matter? Like have X% threshold of being better?
2. Second part checking that we included transactions from at least half of tracked addresses with better transactions than ones in block?

Reply:
That's right. I think an X% threshold is easy to implement if that is what we want. Although, if the best tx is even only slightly better than the best tx in block for 4 blocks straight, that definitely should raise suspicion.